{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5bcb07f8-50bd-4f61-94e1-39e73e5e6922",
   "metadata": {},
   "outputs": [],
   "source": [
    "#=== Install and Imports ===#\n",
    "\n",
    "## installs (in shell) ##\n",
    "## install in a conda env:\n",
    "# conda activate salmon\n",
    "# conda install -c conda-forge -c bioconda salmon\n",
    "\n",
    "\n",
    "## imports ##\n",
    "import pipeline_utils\n",
    "import os\n",
    "import urllib.request\n",
    "from datetime import datetime\n",
    "import sys\n",
    "from contextlib import redirect_stdout\n",
    "import subprocess\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5d00d6c0-169f-4625-a471-565506f780d2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Downloading 1 genomes\n",
      "Currently downloading genome GRCh38\n",
      "Saving genome data at ../0.local/generic-single-cell-pipeline/genomes/human/GRCh38\n",
      "Saving logfile to: ../0.local/generic-single-cell-pipeline/genomes/human/GRCh38/GRCh38_download_log.txt\n"
     ]
    }
   ],
   "source": [
    "#=== Step 1: Generate and document our reference genomes ===#\n",
    "## NOTE: genome assemblies can be obtained at: http://ftp.ensembl.org/pub/\n",
    "## NOTE: check for current builds\n",
    "## NOTE: for documentation purposes, save the date/time and urls in a logfile along with genome files\n",
    "## NOTE: for GRCh37 build;\n",
    "# dna_rm' - masked genomic DNA.  Interspersed repeats and low complexity regions are detected with the RepeatMasker tool and masked by replacing repeats with 'N's.\n",
    "#'dna_sm' - soft-masked genomic DNA. All repeats and low complexity regions have been replaced with lowercased versions of their nucleic base\n",
    "## NOTE: hg19 (maintained by UCSC) = GRCh37 (maintained by genome reference consortium), same for hg38/GRCh38\n",
    "\n",
    "\n",
    "## Define human genome sequences and annotations to download ##\n",
    "human_output_dir = \"../0.local/generic-single-cell-pipeline/genomes/human/\"\n",
    "os.makedirs(human_output_dir, exist_ok=True)\n",
    "human_genomes = {\n",
    "    \"GRCh38\": {\n",
    "        \"genome_url\": \"http://ftp.ensembl.org/pub/release-113/fasta/homo_sapiens/dna/Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz\",\n",
    "        \"gtf_url\": \"http://ftp.ensembl.org/pub/release-113/gtf/homo_sapiens/Homo_sapiens.GRCh38.113.gtf.gz\",\n",
    "        \"cdna_url\": \"http://ftp.ensembl.org/pub/release-113/fasta/homo_sapiens/cdna/Homo_sapiens.GRCh38.cdna.all.fa.gz\",\n",
    "        \"cds_url\": \"http://ftp.ensembl.org/pub/release-113/fasta/homo_sapiens/cds/Homo_sapiens.GRCh38.cds.all.fa.gz\",\n",
    "        \"ncrna_url\": \"http://ftp.ensembl.org/pub/release-113/fasta/homo_sapiens/ncrna/Homo_sapiens.GRCh38.ncrna.fa.gz\",\n",
    "        \"pep_url\": \"http://ftp.ensembl.org/pub/release-113/fasta/homo_sapiens/pep/Homo_sapiens.GRCh38.pep.all.fa.gz\"\n",
    "    }\n",
    "}\n",
    "\n",
    "## Define mouse genome sequences and annotations to download\n",
    "mouse_output_dir = \"../0.local/generic-single-cell-pipeline/genomes/mouse/\"\n",
    "os.makedirs(mouse_output_dir, exist_ok=True)\n",
    "mouse_genomes = {\n",
    "    \"GRCm38\": {\n",
    "        \"fasta_url\": \"http://ftp.ensembl.org/pub/release-111/fasta/mus_musculus/dna/Mus_musculus.GRCm39.dna_sm.primary_assembly.fa.gz\",\n",
    "        \"gtf_url\": \"http://ftp.ensembl.org/pub/release-111/gtf/mus_musculus/Mus_musculus.GRCm39.111.gtf.gz\" \n",
    "     },\n",
    "    \"GRCm39\": {\n",
    "        \"fasta_url\": \"http://ftp.ensembl.org/pub/release-113/fasta/mus_musculus/dna/Mus_musculus.GRCm39.dna.primary_assembly.fa.gz\",\n",
    "        \"gtf_url\": \"http://ftp.ensembl.org/pub/release-113/gtf/mus_musculus/Mus_musculus.GRCm39.113.gtf.gz\"\n",
    "    }\n",
    "}\n",
    "\n",
    "#### Download\n",
    "pipeline_utils.download_genome(human_genomes, human_output_dir)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "b558a9f6-66d5-4895-84a5-26a9acf44c6d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Version Info: This is the most recent version of salmon.\n",
      "[2025-04-18 13:15:15.496] [jLog] [warning] The salmon index is being built without any decoy sequences.  It is recommended that decoy sequence (either computed auxiliary decoy sequence or the genome of the organism) be provided during indexing. Further details can be found at https://salmon.readthedocs.io/en/latest/salmon.html#preparing-transcriptome-indices-mapping-based-mode.\n",
      "[2025-04-18 13:15:15.496] [jLog] [info] building index\n",
      "out : /home/ubuntu1/GitHub/0.local/generic-single-cell-pipeline/genomes/human/GRCh38/index/transcript/salmon/\n",
      "[2025-04-18 13:15:15.496] [puff::index::jointLog] [info] Running fixFasta\n",
      "\n",
      "[Step 1 of 4] : counting k-mers\n",
      "[2025-04-18 13:15:15.516] [puff::index::jointLog] [warning] Entry with header [ENST00000415118.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:15.516] [puff::index::jointLog] [warning] Entry with header [ENST00000448914.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:15.516] [puff::index::jointLog] [warning] Entry with header [ENST00000434970.2], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:15.516] [puff::index::jointLog] [warning] Entry with header [ENST00000632684.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:15.516] [puff::index::jointLog] [warning] Entry with header [ENST00000631435.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:15.516] [puff::index::jointLog] [warning] Entry with header [ENST00000710614.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:15.523] [puff::index::jointLog] [warning] Entry with header [ENST00000632524.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:15.523] [puff::index::jointLog] [warning] Entry with header [ENST00000633009.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:15.524] [puff::index::jointLog] [warning] Entry with header [ENST00000634070.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:15.524] [puff::index::jointLog] [warning] Entry with header [ENST00000632963.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:15.524] [puff::index::jointLog] [warning] Entry with header [ENST00000633030.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:15.524] [puff::index::jointLog] [warning] Entry with header [ENST00000633765.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:15.524] [puff::index::jointLog] [warning] Entry with header [ENST00000632619.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:15.524] [puff::index::jointLog] [warning] Entry with header [ENST00000632968.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:15.524] [puff::index::jointLog] [warning] Entry with header [ENST00000633159.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:15.524] [puff::index::jointLog] [warning] Entry with header [ENST00000631871.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:15.524] [puff::index::jointLog] [warning] Entry with header [ENST00000633010.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:15.524] [puff::index::jointLog] [warning] Entry with header [ENST00000632473.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:15.524] [puff::index::jointLog] [warning] Entry with header [ENST00000631884.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:15.524] [puff::index::jointLog] [warning] Entry with header [ENST00000632859.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:15.524] [puff::index::jointLog] [warning] Entry with header [ENST00000631895.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:15.524] [puff::index::jointLog] [warning] Entry with header [ENST00000634154.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:15.524] [puff::index::jointLog] [warning] Entry with header [ENST00000632609.2], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:15.524] [puff::index::jointLog] [warning] Entry with header [ENST00000632911.2], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:15.524] [puff::index::jointLog] [warning] Entry with header [ENST00000633504.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:15.524] [puff::index::jointLog] [warning] Entry with header [ENST00000632304.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:15.524] [puff::index::jointLog] [warning] Entry with header [ENST00000632542.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:15.524] [puff::index::jointLog] [warning] Entry with header [ENST00000633968.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:15.524] [puff::index::jointLog] [warning] Entry with header [ENST00000634085.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:15.524] [puff::index::jointLog] [warning] Entry with header [ENST00000633353.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:15.524] [puff::index::jointLog] [warning] Entry with header [ENST00000631803.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:15.524] [puff::index::jointLog] [warning] Entry with header [ENST00000633210.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:15.524] [puff::index::jointLog] [warning] Entry with header [ENST00000439842.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:15.524] [puff::index::jointLog] [warning] Entry with header [ENST00000390567.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:15.524] [puff::index::jointLog] [warning] Entry with header [ENST00000452198.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:15.524] [puff::index::jointLog] [warning] Entry with header [ENST00000390569.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:15.524] [puff::index::jointLog] [warning] Entry with header [ENST00000437320.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:15.524] [puff::index::jointLog] [warning] Entry with header [ENST00000390571.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:15.524] [puff::index::jointLog] [warning] Entry with header [ENST00000390572.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:15.524] [puff::index::jointLog] [warning] Entry with header [ENST00000450276.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:15.524] [puff::index::jointLog] [warning] Entry with header [ENST00000390574.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:15.524] [puff::index::jointLog] [warning] Entry with header [ENST00000390575.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:15.524] [puff::index::jointLog] [warning] Entry with header [ENST00000431870.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:15.524] [puff::index::jointLog] [warning] Entry with header [ENST00000390578.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:15.524] [puff::index::jointLog] [warning] Entry with header [ENST00000451044.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:15.524] [puff::index::jointLog] [warning] Entry with header [ENST00000390580.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:15.524] [puff::index::jointLog] [warning] Entry with header [ENST00000390581.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:15.524] [puff::index::jointLog] [warning] Entry with header [ENST00000431440.2], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:15.524] [puff::index::jointLog] [warning] Entry with header [ENST00000390585.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:15.524] [puff::index::jointLog] [warning] Entry with header [ENST00000430425.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:15.524] [puff::index::jointLog] [warning] Entry with header [ENST00000454691.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:15.524] [puff::index::jointLog] [warning] Entry with header [ENST00000390588.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:15.524] [puff::index::jointLog] [warning] Entry with header [ENST00000414852.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:15.524] [puff::index::jointLog] [warning] Entry with header [ENST00000390590.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:15.524] [puff::index::jointLog] [warning] Entry with header [ENST00000390591.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:15.524] [puff::index::jointLog] [warning] Entry with header [ENST00000454908.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:15.524] [puff::index::jointLog] [warning] Entry with header [ENST00000390583.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:15.524] [puff::index::jointLog] [warning] Entry with header [ENST00000390584.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:15.524] [puff::index::jointLog] [warning] Entry with header [ENST00000604642.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:15.524] [puff::index::jointLog] [warning] Entry with header [ENST00000603326.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:15.524] [puff::index::jointLog] [warning] Entry with header [ENST00000604950.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:15.524] [puff::index::jointLog] [warning] Entry with header [ENST00000603077.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:15.524] [puff::index::jointLog] [warning] Entry with header [ENST00000605284.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:15.524] [puff::index::jointLog] [warning] Entry with header [ENST00000604446.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:15.524] [puff::index::jointLog] [warning] Entry with header [ENST00000603693.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:15.524] [puff::index::jointLog] [warning] Entry with header [ENST00000603935.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:15.524] [puff::index::jointLog] [warning] Entry with header [ENST00000604102.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:15.524] [puff::index::jointLog] [warning] Entry with header [ENST00000604838.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[2025-04-18 13:15:18.208] [puff::index::jointLog] [warning] Entry with header [ENST00000626807.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:18.482] [puff::index::jointLog] [warning] Entry with header [ENST00000629250.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:18.511] [puff::index::jointLog] [warning] Entry with header [ENST00000632054.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:23.794] [puff::index::jointLog] [warning] Entry with header [ENST00000437226.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:24.609] [puff::index::jointLog] [warning] Entry with header [ENST00000417283.5], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:24.609] [puff::index::jointLog] [warning] Entry with header [ENST00000465958.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:25.623] [puff::index::jointLog] [warning] Entry with header [ENST00000692460.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:29.770] [puff::index::jointLog] [warning] Entry with header [ENST00000682202.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:33.086] [puff::index::jointLog] [warning] Entry with header [ENST00000473810.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:33.101] [puff::index::jointLog] [warning] Entry with header [ENST00000543745.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:33.172] [puff::index::jointLog] [warning] Entry with header [ENST00000603775.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:33.202] [puff::index::jointLog] [warning] Entry with header [ENST00000579054.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:33.525] [puff::index::jointLog] [warning] Entry with header [ENST00000634174.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:33.543] [puff::index::jointLog] [warning] Entry with header [ENST00000573437.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:33.668] [puff::index::jointLog] [warning] Entry with header [ENST00000518246.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "[2025-04-18 13:15:33.669] [puff::index::jointLog] [warning] Entry with header [ENST00000632342.1], had length less than equal to the k-mer length of 31 (perhaps after poly-A clipping)\n",
      "\n",
      "[2025-04-18 13:15:33.745] [puff::index::jointLog] [warning] Removed 12966 transcripts that were sequence duplicates of indexed transcripts.\n",
      "[2025-04-18 13:15:33.745] [puff::index::jointLog] [warning] If you wish to retain duplicate transcripts, please use the `--keepDuplicates` flag\n",
      "[2025-04-18 13:15:33.753] [puff::index::jointLog] [info] Replaced 100,005 non-ATCG nucleotides\n",
      "[2025-04-18 13:15:33.753] [puff::index::jointLog] [info] Clipped poly-A tails from 1,517 transcripts\n",
      "wrote 194120 cleaned references\n",
      "[2025-04-18 13:15:35.100] [puff::index::jointLog] [info] Filter size not provided; estimating from number of distinct k-mers\n",
      "[2025-04-18 13:15:41.062] [puff::index::jointLog] [info] ntHll estimated 115908649 distinct k-mers, setting filter size to 2^31\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Threads = 16\n",
      "Vertex length = 31\n",
      "Hash functions = 5\n",
      "Filter size = 2147483648\n",
      "Capacity = 2\n",
      "Files: \n",
      "/home/ubuntu1/GitHub/0.local/generic-single-cell-pipeline/genomes/human/GRCh38/index/transcript/salmon/ref_k31_fixed.fa\n",
      "--------------------------------------------------------------------------------\n",
      "Round 0, 0:2147483648\n",
      "Pass\tFilling\tFiltering\n",
      "1\t14\t22\t\n",
      "2\t17\t0\n",
      "True junctions count = 796453\n",
      "False junctions count = 1438281\n",
      "Hash table size = 2234734\n",
      "Candidate marks count = 13034038\n",
      "--------------------------------------------------------------------------------\n",
      "Reallocating bifurcations time: 1\n",
      "True marks count: 8240193\n",
      "Edges construction time: 17\n",
      "--------------------------------------------------------------------------------\n",
      "Distinct junctions = 796453\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TwoPaCo::buildGraphMain:: allocated with scalable_malloc; freeing.\n",
      "TwoPaCo::buildGraphMain:: Calling scalable_allocation_command(TBBMALLOC_CLEAN_ALL_BUFFERS, 0);\n",
      "allowedIn: 18\n",
      "Max Junction ID: 903709\n",
      "seen.size():7229681 kmerInfo.size():903710\n",
      "approximateContigTotalLength: 81004901\n",
      "counters for complex kmers:\n",
      "(prec>1 & succ>1)=55259 | (succ>1 & isStart)=849 | (prec>1 & isEnd)=854 | (isStart & isEnd)=80\n",
      "contig count: 1195650 element count: 152578102 complex nodes: 57042\n",
      "# of ones in rank vector: 1195649\n",
      "[2025-04-18 13:17:23.022] [puff::index::jointLog] [info] Starting the Pufferfish indexing by reading the GFA binary file.\n",
      "[2025-04-18 13:17:23.022] [puff::index::jointLog] [info] Setting the index/BinaryGfa directory /home/ubuntu1/GitHub/0.local/generic-single-cell-pipeline/genomes/human/GRCh38/index/transcript/salmon\n",
      "size = 152578102\n",
      "-----------------------------------------\n",
      "| Loading contigs | Time = 26.286 ms\n",
      "-----------------------------------------\n",
      "size = 152578102\n",
      "-----------------------------------------\n",
      "| Loading contig boundaries | Time = 13.049 ms\n",
      "-----------------------------------------\n",
      "Number of ones: 1195649\n",
      "Number of ones per inventory item: 512\n",
      "Inventory entries filled: 2336\n",
      "1195649\n",
      "[2025-04-18 13:17:23.464] [puff::index::jointLog] [info] Done wrapping the rank vector with a rank9sel structure.\n",
      "[2025-04-18 13:17:23.478] [puff::index::jointLog] [info] contig count for validation: 1,195,649\n",
      "[2025-04-18 13:17:24.122] [puff::index::jointLog] [info] Total # of Contigs : 1,195,649\n",
      "[2025-04-18 13:17:24.122] [puff::index::jointLog] [info] Total # of numerical Contigs : 1,195,649\n",
      "[2025-04-18 13:17:24.229] [puff::index::jointLog] [info] Total # of contig vec entries: 8,296,889\n",
      "[2025-04-18 13:17:24.229] [puff::index::jointLog] [info] bits per offset entry 23\n",
      "[2025-04-18 13:17:24.710] [puff::index::jointLog] [info] Done constructing the contig vector. 1195650\n",
      "[2025-04-18 13:17:25.229] [puff::index::jointLog] [info] # segments = 1,195,649\n",
      "[2025-04-18 13:17:25.229] [puff::index::jointLog] [info] total length = 152,578,102\n",
      "[2025-04-18 13:17:25.325] [puff::index::jointLog] [info] Reading the reference files ...\n",
      "[2025-04-18 13:17:26.868] [puff::index::jointLog] [info] positional integer width = 28\n",
      "[2025-04-18 13:17:26.868] [puff::index::jointLog] [info] seqSize = 152,578,102\n",
      "[2025-04-18 13:17:26.868] [puff::index::jointLog] [info] rankSize = 152,578,102\n",
      "[2025-04-18 13:17:26.869] [puff::index::jointLog] [info] edgeVecSize = 0\n",
      "[2025-04-18 13:17:26.869] [puff::index::jointLog] [info] num keys = 116,708,632\n",
      "[Building BooPHF]  100  %   elapsed:   0 min 6  sec   remaining:   0 min 0  sec\n",
      "[2025-04-18 13:17:32.549] [puff::index::jointLog] [info] mphf size = 72.899 MB\n",
      "[2025-04-18 13:17:32.854] [puff::index::jointLog] [info] chunk size = 9,536,132\n",
      "[2025-04-18 13:17:32.854] [puff::index::jointLog] [info] chunk 0 = [0, 9,536,132)\n",
      "[2025-04-18 13:17:32.854] [puff::index::jointLog] [info] chunk 1 = [9,536,132, 19,072,269)\n",
      "[2025-04-18 13:17:32.854] [puff::index::jointLog] [info] chunk 2 = [19,072,269, 28,608,401)\n",
      "[2025-04-18 13:17:32.854] [puff::index::jointLog] [info] chunk 3 = [28,608,401, 38,144,545)\n",
      "[2025-04-18 13:17:32.854] [puff::index::jointLog] [info] chunk 4 = [38,144,545, 47,680,677)\n",
      "[2025-04-18 13:17:32.854] [puff::index::jointLog] [info] chunk 5 = [47,680,677, 57,216,809)\n",
      "[2025-04-18 13:17:32.854] [puff::index::jointLog] [info] chunk 6 = [57,216,809, 66,752,941)\n",
      "[2025-04-18 13:17:32.854] [puff::index::jointLog] [info] chunk 7 = [66,752,941, 76,289,073)\n",
      "[2025-04-18 13:17:32.854] [puff::index::jointLog] [info] chunk 8 = [76,289,073, 85,825,220)\n",
      "[2025-04-18 13:17:32.854] [puff::index::jointLog] [info] chunk 9 = [85,825,220, 95,361,352)\n",
      "[2025-04-18 13:17:32.854] [puff::index::jointLog] [info] chunk 10 = [95,361,352, 104,897,484)\n",
      "[2025-04-18 13:17:32.854] [puff::index::jointLog] [info] chunk 11 = [104,897,484, 114,433,616)\n",
      "[2025-04-18 13:17:32.854] [puff::index::jointLog] [info] chunk 12 = [114,433,616, 123,969,771)\n",
      "[2025-04-18 13:17:32.854] [puff::index::jointLog] [info] chunk 13 = [123,969,771, 133,505,903)\n",
      "[2025-04-18 13:17:32.854] [puff::index::jointLog] [info] chunk 14 = [133,505,903, 143,042,035)\n",
      "[2025-04-18 13:17:32.854] [puff::index::jointLog] [info] chunk 15 = [143,042,035, 152,578,072)\n",
      "[2025-04-18 13:17:37.180] [puff::index::jointLog] [info] finished populating pos vector\n",
      "[2025-04-18 13:17:37.180] [puff::index::jointLog] [info] writing index components\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "for info, total work write each  : 2.331    total work inram from level 3 : 4.322  total work raw : 25.000 \n",
      "Bitarray       611521472  bits (100.00 %)   (array + ranks )\n",
      "final hash             0  bits (0.00 %) (nb in final hash 0)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[2025-04-18 13:17:37.666] [puff::index::jointLog] [info] finished writing dense pufferfish index\n",
      "[2025-04-18 13:17:37.825] [jLog] [info] done building index\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "CompletedProcess(args=['salmon', 'index', '-t', '/home/ubuntu1/GitHub/0.local/generic-single-cell-pipeline/genomes/human/GRCh38/Homo_sapiens.GRCh38.cdna.all.fa.gz', '-i', '/home/ubuntu1/GitHub/0.local/generic-single-cell-pipeline/genomes/human/GRCh38/index/transcript/salmon/', '--threads', '16'], returncode=0)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Salmon, like eXpress 1, uses a streaming inference method to perform transcript-level quantification.\n",
    "# One of the fundamental assumptions of such inference methods is that observations (i.e. reads or alignments)\n",
    "# are made “at random”. This means, for example, that alignments should not be sorted by target or position.\n",
    "# If your reads or alignments do not appear in a random order with respect to the target transcripts,\n",
    "# please randomize / shuffle them before performing quantification with Salmon.\n",
    "\n",
    "#=== Build transcript indexes using salmon ===#\n",
    "cdna_path = \"/home/ubuntu1/GitHub/0.local/generic-single-cell-pipeline/genomes/human/GRCh38/Homo_sapiens.GRCh38.cdna.all.fa.gz\"\n",
    "index_path = \"/home/ubuntu1/GitHub/0.local/generic-single-cell-pipeline/genomes/human/GRCh38/index/transcript/salmon/\"\n",
    "#logfile_path = \"index/transcript/salmon/nohup_log.txt\"\n",
    "#working_dir = \"/home/ubuntu1/GitHub/0.local/generic-single-cell-pipeline/genomes/human/GRCh38/\"\n",
    "\n",
    "\n",
    "subprocess.run([\n",
    "    \"salmon\", \"index\",\n",
    "    \"-t\", cdna_path,\n",
    "    \"-i\", index_path,\n",
    "    \"--threads\", \"16\"\n",
    "])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "dc14d46d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "refdata-gex-GRCh38-2024-A/\n",
      "refdata-gex-GRCh38-2024-A/reference.json\n",
      "refdata-gex-GRCh38-2024-A/star/\n",
      "refdata-gex-GRCh38-2024-A/star/exonGeTrInfo.tab\n",
      "refdata-gex-GRCh38-2024-A/star/sjdbInfo.txt\n",
      "refdata-gex-GRCh38-2024-A/star/chrLength.txt\n",
      "refdata-gex-GRCh38-2024-A/star/exonInfo.tab\n",
      "refdata-gex-GRCh38-2024-A/star/Genome\n",
      "refdata-gex-GRCh38-2024-A/star/chrName.txt\n",
      "refdata-gex-GRCh38-2024-A/star/chrStart.txt\n",
      "refdata-gex-GRCh38-2024-A/star/chrNameLength.txt\n",
      "refdata-gex-GRCh38-2024-A/star/geneInfo.tab\n",
      "refdata-gex-GRCh38-2024-A/star/transcriptInfo.tab\n",
      "refdata-gex-GRCh38-2024-A/star/sjdbList.fromGTF.out.tab\n",
      "refdata-gex-GRCh38-2024-A/star/sjdbList.out.tab\n",
      "refdata-gex-GRCh38-2024-A/star/genomeParameters.txt\n",
      "refdata-gex-GRCh38-2024-A/star/SA\n",
      "refdata-gex-GRCh38-2024-A/star/SAindex\n",
      "refdata-gex-GRCh38-2024-A/fasta/\n",
      "refdata-gex-GRCh38-2024-A/fasta/genome.fa.fai\n",
      "refdata-gex-GRCh38-2024-A/fasta/genome.fa\n",
      "refdata-gex-GRCh38-2024-A/genes/\n",
      "refdata-gex-GRCh38-2024-A/genes/genes.gtf.gz\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "CompletedProcess(args=['tar', '-xvzf', 'refdata-gex-GRCh38-2024-A.tar.gz'], returncode=0)"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#=== Processing single-cell data generated by 10X with cell ranger ===#\n",
    "# see link for guide: https://www.10xgenomics.com/support/software/cell-ranger/downloads#download-links\n",
    "\n",
    "## get the compiled binary file ##\n",
    "#url = \"https://cf.10xgenomics.com/releases/cell-exp/cellranger-9.0.1.tar.gz?Expires=1745041256&Key-Pair-Id=APKAI7S6A5RYOXBWRPDA&Signature=JIIo9YRo~c4N0qV9Jk~pf0bIijj7E4wYUbHbcyVZEhj0lDIgrVsXSzV2623rrjdlZ2EVM8M5R-1Kaej5HsAtDy2Ib~Jjr9YlSvRMO0U5mEDo9cYK0kd5jZFpRkDs2QByTEb0iE-5MFAIqq2yTLnyyjp4xGQB1UTOQ2M2romVOBLBZpZ2ulx0Myg-KcSWxTsbmzVuhM7YW~wYKJ22s~R1sQl9bJ~ybQTa-eoCL1x7LjFlfGdUzKQPjapTBvhzqU0xkPIhHWiiqjKRSZmEDDQKAcL4y~P~WauLdUP6VqTBRAtdW4NcZ6wE03t67M~nVGVffE6LWsuVFFSFdv43xUV-YQ__\"\n",
    "#\n",
    "#subprocess.run([\"wget\", \"-O\", \"cellranger-9.0.1.tar.gz\", url],\n",
    "#              cwd=\"/home/ubuntu1/programs/cellranger/\")\n",
    "\n",
    "# untar gunzip\n",
    "#subprocess.run([\"tar\", \"-xvzf\", \"cellranger-9.0.1.tar.gz\"],\n",
    "#              cwd=\"/home/ubuntu1/programs/cellranger/\")\n",
    "\n",
    "# add to path variable, will only work for interactive shell\n",
    "#cellranger_path = \"/home/ubuntu1/programs/cellranger/cellranger-9.0.1/bin/\"\n",
    "#bashrc_line = f'\\nexport PATH=\"{cellranger_path}:$PATH\"\\n'\n",
    "#subprocess.run(\n",
    "#    f'echo \"{bashrc_line}\" >> ~/.bashrc',\n",
    "#    shell=True,\n",
    "#    executable=\"/bin/bash\"\n",
    "#)\n",
    "\n",
    "# check the path in python\n",
    "#import os\n",
    "#print(os.environ[\"PATH\"])\n",
    "\n",
    "## NOTE: it is required to explicitly specify the path of cellranger here to run the subprocess command ##\n",
    "# test install\n",
    "#env = os.environ.copy()\n",
    "#env[\"PATH\"] = f\"/home/ubuntu1/programs/cellranger/cellranger-9.0.1/bin/:{env['PATH']}\"\n",
    "#subprocess.run([\"cellranger\", \"--version\"], env=env)\n",
    "\n",
    "## download the reference genome ##\n",
    "#url = \"https://cf.10xgenomics.com/supp/cell-exp/refdata-gex-GRCh38-2024-A.tar.gz\"\n",
    "#subprocess.run([\"mkdir\", \"reference\"],\n",
    "              #cwd=\"/home/ubuntu1/programs/cellranger/\")\n",
    "#subprocess.run([\"wget\", \"-O\", \"refdata-gex-GRCh38-2024-A.tar.gz\", url],\n",
    "              #cwd=\"/home/ubuntu1/programs/cellranger/reference\")\n",
    "\n",
    "# untar gunzip\n",
    "#subprocess.run([\"tar\", \"-xvzf\", \"refdata-gex-GRCh38-2024-A.tar.gz\"],\n",
    "#              cwd=\"/home/ubuntu1/programs/cellranger/reference\")\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8893993d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#=== Find sequencing runs that use 3' capture technology ===#\n",
    "\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
